src/librustdoc/html/static/js/search.js

```diff
@@ -133,18 +124,271 @@ window.initSearch = function(rawSearchIndex) {
+    }
+    function removeEmptyStringsFromArray(arr) {
+        for (var i = 0, len = arr.length; i < len; ++i) {
+            if (arr[i] === "") {
+                arr.splice(i, 1);
+                i -= 1;
+            }
+        }
+    }
+
     /**
-     * Executes the query and returns a list of results for each results tab.
-     * @param  {Object}        query          - The user query
-     * @param  {Array<string>} searchWords    - The list of search words to query against
-     * @param  {string}        [filterCrates] - Crate to search in
-     * @return {{
-     *   in_args: Array<?>,
-     *   returned: Array<?>,
-     *   others: Array<?>
-     * }}
+     * Executes the query and builds an index of results
+     *
+     * The supported syntax by this parser is as follow:
+     *
+     * ident = *1(ALPHA / DIGIT)
+     * path = ident *WS *(DOUBLE-COLON *WS ident)
+     * arg = path *WS [generics]
+     * nonempty-arg-list = arg *WS *(COMMA *WS arg)
+     * generics = OPEN-ANGLE-BRACKET *WS nonempty-arg-list *WS CLOSE-ANGLE-BRACKET
+     * function-args = OPEN-PAREN *WS [nonempty-arg-list] *WS END-PAREN
+     * return-args = RETURN-ARROW *WS function-args
+     *
+     * exact-search = [type-filter *WS COLON] *WS QUOTE ident QUOTE *WS [generics]
+     * type-search = [type-filter *WS COLON] *WS path *WS generics
+     * function-search = path *WS function-args *WS [return-args]
+     *
+     * query = *WS (exact-search / type-search / function-search / return-args) *WS
+     *
+     * type-filter = (
+     *     "mod" /
+     *     "externcrate" /
+     *     "import" /
+     *     "struct" /
+     *     "enum" /
+     *     "fn" /
+     *     "type" /
+     *     "static" /
+     *     "trait" /
+     *     "impl" /
+     *     "tymethod" /
+     *     "method" /
+     *     "structfield" /
+     *     "variant" /
+     *     "macro" /
+     *     "primitive" /
+     *     "associatedtype" /
+     *     "constant" /
+     *     "associatedconstant" /
+     *     "union" /
+     *     "foreigntype" /
+     *     "keyword" /
+     *     "existential" /
+     *     "attr" /
+     *     "derive" /
+     *     "traitalias")
+     *
+     * OPEN-ANGLE-BRACKET = "<"
+     * CLOSE-ANGLE-BRACKET = ">"
+     * OPEN-PAREN = "("
+     * END-PAREN = ")"
+     * COLON = ":"
+     * DOUBLE-COLON = "::"
+     * QUOTE = %x22
+     * COMMA = ","
+     * RETURN-ARROW = "->"
+     *
+     * ALPHA = %x41-5A / %x61-7A ; A-Z / a-z
+     * DIGIT = %x30-39
+     * WS = %x09 / " "
+     *
+     * @param  {string} val     - The user query
+     * @return {ParsedQuery}    - The parsed query
      */
-    function execQuery(query, searchWords, filterCrates) {
+    function parseQuery(userQuery) {
+        var queryLength = userQuery.length;
+        var queryPos = 0;
+        // Total number of elements (includes generics).
+        var totalElems = 0;
+
+        function getStringElem(query, isInGenerics) {
```

Comment:
I see you move isWhitespace, isStopCharacter, etc. outside of `parseQuery`. But there are still a lot of nested functions in `parseQuery`. I'd like you to move _all_ of them out.